extension/src/datum_utils.rs

```diff
@@ -198,55 +198,98 @@ flat_serialize_macro::flat_serialize! {
+            pg_sys::fmgr_info(type_output, &mut flinfo);
+        }
+        let count = self.iter().count();
+        let mut seq = serializer.serialize_seq(Some(count + 1))?;
+        seq.serialize_element(&self.type_oid.0)?;
+        for element in self.iter() {
+            let chars = unsafe{pg_sys::OutputFunctionCall(&mut flinfo, element)};
+            let cstr = unsafe {
+                std::ffi::CStr::from_ptr(chars)
+            };
+            seq.serialize_element(cstr.to_str().unwrap())?;
+        }
+        seq.end()
+    }
+}
+
+impl<'a, 'de> Deserialize<'de> for DatumStore<'a> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>
+    {
+        struct DatumStoreVisitor<'a>(std::marker::PhantomData<core::cell::Cell<&'a ()>>);
+
+        impl<'de, 'a> Visitor<'de> for DatumStoreVisitor<'a> {
+            type Value = DatumStore<'a>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a sequence encoding a DatumStore object")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                // borrowed from serialization/types.rs
+                unsafe fn get_struct<T>(tuple: pg_sys::HeapTuple) -> *mut T {
+                    //((char *) ((TUP)->t_data) + (TUP)->t_data->t_hoff)
+                    let t_data: *mut u8 = (*tuple).t_data.cast();
+                    let t_hoff = (*(*tuple).t_data).t_hoff;
+                    t_data.add(t_hoff as usize).cast()
+                }
+
+                let oid = seq.next_element::<Oid>().unwrap().unwrap();  // TODO: error handling
+
+                let mut type_input = 0;
+                let mut typ_io_param = 0;
+                let mut flinfo = unsafe {
+                    std::mem::MaybeUninit::zeroed().assume_init()
+                };
+                let typmod;
+                unsafe {
+                    pg_sys::getTypeInputInfo(oid, &mut type_input, &mut typ_io_param);
+                    pg_sys::fmgr_info(type_input, &mut flinfo);
+
+                    // Is there an easier way to get the typmod...is this even getting the right value?
+                    let tuple =
+                        pg_sys::SearchSysCache1(pg_sys::SysCacheIdentifier_TYPEOID as _, oid as _);
+                    if tuple.is_null() {
+                        pgx::error!("no type info for oid {}", oid);
+                    }
+        
+                    let type_tuple: pg_sys::Form_pg_type = get_struct(tuple);
+                    typmod = (*type_tuple).typmodin;
+                    pg_sys::ReleaseSysCache(tuple);
+                }
+
+                let mut data = vec![];
+                while let Some(next) = seq.next_element::<String>()? {
+                    let next = std::ffi::CString::new(next).unwrap();  // TODO: error handling
```

Comment:
annoying that pg forces an extra allocation here, but them's the breaks

Reply:
Yeah, I was annoyed with this as well.